File: apps/web/src/lib/api/mood.ts
import { apiFetch } from "./client";

export type InputMethod = "tap" | "voice" | "text";

export interface MoodCreateRequest {
  mood_score: number;
  energy_score?: number | null;
  stress_score?: number | null;
  emotion_tags?: string[];
  context_tag?: string;
  note?: string;
  input_method?: InputMethod;
}

export interface MoodResponse {
  id: string;
  mood_score: number;
  energy_score: number | null;
  stress_score: number | null;
  emotion_tags: string[];
  context_tag: string | null;
  note: string | null;
  input_method: InputMethod;
  logged_at: string;
  created_at: string;
}

export interface MoodAnalyticsBucket {
  period: string;
  period_start: string;
  entries: number;
  avg_mood: number | null;
  avg_energy: number | null;
  avg_stress: number | null;
}

export interface EmotionCount {
  emotion: string;
  count: number;
}

export interface MoodAnalytics {
  period: "weekly" | "monthly";
  range_days: number;
  total_entries: number;
  avg_mood: number | null;
  avg_energy: number | null;
  avg_stress: number | null;
  buckets: MoodAnalyticsBucket[];
  top_emotions: EmotionCount[];
}

export async function submitMoodLog(data: MoodCreateRequest): Promise<MoodResponse> {
  const res = await apiFetch("/mood", {
    method: "POST",
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error("Failed to submit mood log");
  return res.json() as Promise<MoodResponse>;
}

export async function getMoodHistory(days: number = 30): Promise<MoodResponse[]> {
  const res = await apiFetch(`/mood/history?days=${days}`);
  if (!res.ok) throw new Error("Failed to get mood history");
  return res.json() as Promise<MoodResponse[]>;
}

export async function getMoodAnalytics(
  period: "weekly" | "monthly" = "weekly",
  days = 90,
): Promise<MoodAnalytics> {
  const res = await apiFetch(`/mood/analytics?period=${period}&days=${days}`);
  if (!res.ok) throw new Error("Failed to get mood analytics");
  return res.json() as Promise<MoodAnalytics>;
}

File: apps/api/app/schemas/mood.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.mood_log import InputMethod


class MoodCreateRequest(BaseModel):
    mood_score: int = Field(..., ge=1, le=5)
    energy_score: int | None = Field(default=None, ge=1, le=5)
    stress_score: int | None = Field(default=None, ge=1, le=5)
    emotion_tags: list[str] = Field(default_factory=list, max_length=15)
    context_tag: str | None = Field(default=None, max_length=50)
    note: str | None = Field(default=None, max_length=1000)
    input_method: InputMethod = Field(default=InputMethod.TAP)

    model_config = ConfigDict(extra="forbid")


class MoodResponse(BaseModel):
    id: uuid.UUID
    mood_score: int
    energy_score: int | None
    stress_score: int | None
    emotion_tags: list[str]
    context_tag: str | None
    note: str | None
    input_method: InputMethod
    logged_at: datetime
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class MoodAnalyticsBucket(BaseModel):
    """One period (week or month) of aggregated mood data."""

    period: str  # e.g. "2026-W29" for a week, or "2026-07" for a month
    period_start: str  # ISO date of the first day in the period
    entries: int
    avg_mood: float | None
    avg_energy: float | None
    avg_stress: float | None


class EmotionCount(BaseModel):
    emotion: str
    count: int


class MoodAnalyticsResponse(BaseModel):
    period: str  # "weekly" or "monthly"
    range_days: int
    total_entries: int
    avg_mood: float | None
    avg_energy: float | None
    avg_stress: float | None
    buckets: list[MoodAnalyticsBucket]
    top_emotions: list[EmotionCount]
